src/wasm/wasm-validator.cpp

```diff
@@ -2446,7 +2446,79 @@ void FunctionValidator::visitTry(Try* curr) {
 }
 
 void FunctionValidator::visitTryTable(TryTable* curr) {
-  // TODO
+  shouldBeTrue(
+    getModule()->features.hasExceptionHandling(),
+    curr,
+    "try_table requires exception-handling [--enable-exception-handling]");
+  if (curr->type != Type::unreachable) {
+    shouldBeSubType(curr->body->type,
+                    curr->type,
+                    curr->body,
+                    "try_table's type does not match try_table body's type");
+  } else {
+    shouldBeEqual(curr->body->type,
+                  Type(Type::unreachable),
+                  curr,
+                  "unreachable try_table must have unreachable try_table body");
+  }
+
+  shouldBeEqual(curr->catchTags.size(),
+                curr->catchDests.size(),
+                curr,
+                "the number of catch tags and catch destinations do not match");
+  shouldBeEqual(curr->catchTags.size(),
+                curr->catchRefs.size(),
+                curr,
+                "the number of catch tags and catch refs do not match");
+  shouldBeEqual(curr->catchTags.size(),
+                curr->sentTypes.size(),
+                curr,
+                "the number of catch tags and sent types do not match");
+
+  Type exnref = Type(HeapType::exn, Nullable);
+  for (Index i = 0; i < curr->catchTags.size(); i++) {
+    Name tagName = curr->catchTags[i];
+    if (!tagName) {
+      continue;
+    }
+
+    // Check tag validity
+    auto* tag = getModule()->getTagOrNull(tagName);
+    if (!shouldBeTrue(tag != nullptr, curr, "")) {
+      getStream() << "catch's tag name is invalid: " << tagName << "\n";
+    } else if (!shouldBeEqual(tag->sig.results, Type(Type::none), curr, "")) {
+      getStream()
+        << "catch's tag (" << tagName
+        << ") has result values, which is not allowed for exception handling";
+    }
+
+    // Check types in sentTypes is valid
+    auto tagType = tag->sig.params;
+    auto sentType = curr->sentTypes[i];
+    if (tagType.size() == 0 && !curr->catchRefs[i]) {
+      if (shouldBeTrue(sentType.size() == 1, curr, "")) {
+        shouldBeEqual(sentType,
+                      Type(Type::none),
+                      curr,
+                      "invalid catch sent type information");
+      }
+    }
+    for (Index j = 0; j < tagType.size(); j++) {
+      shouldBeEqual(
+        tagType[j], sentType[j], curr, "invalid catch sent type information");
+    }
+    if (curr->catchRefs[i]) {
```

Comment:
It would be good to check that the sizes match when this is false as well. Alternatively, when there is no extra exnref, you can just check `tagType == sentType`.

Reply:
This was actually incorrect because even if there is no values to send, `sentType` will contain `Type::none`, which is size 1... And this wasn't crashing because I was doing the early exit (https://github.com/WebAssembly/binaryen/pull/6185/files#diff-2149be365b66a82038f1d3fa8f9fb1b4dcd40ab535c986f4bf0a4c37e669a1ceR2481-R2483). Will fix that and check both cases.

Reply:
Nevermind the comment. Didn't know `(Type::none).size()` returns 0...

Reply:
Ended up rewriting much of the loop, but I think I an now checking both cases. PTAL